### ISSUE TITLE
Is my fine-tuned sentence transformer overfitted?

### ISSUE THREAD

**Issue by vahuja4** [huggingface/setfit#406]
I have a text classification problem with around 130 classes. I am using e5-base-v2 as the body and fine-tuning for 20 epochs. Anything less lowers the performance and anything more doesn't help. But what I have found is that no matter which scikit head I use (LR, RF, Gradient Boosting, Extra Trees, SVM), my final performance remains the same. Does this mean that the sentence transformer is separating the embeddings too well leaving very little work for the classifier to do?

**Comment by kgourgou** [comment]
> Does this mean that the sentence transformer is separating the embeddings too well leaving very little work for the classifier to do?

This is a great question and  I think your hypothesis may be reasonable. 20 epochs is definitely way more than I've tried and sounds likely that you have overfitted. 

One way to see if that is the case is to try a TSNE plot to visualize the embeddings. The performance of a nearest neighbor classifier on the embeddings may also give a hint.

**Comment by tomaarsen** [comment]
I'm working on a PR to introduce some semblance of logging during training (#265). It includes training and evaluation losses, which could indicate whether the model is overfitting (i.e. training loss at 0 while evaluation loss grows). I'd like to get it merged soon, but I think it might still take a while I'm afraid.

The TSNE idea from @kgourgou is great too. 

P.s. thanks for always taking the time to respond to questions here @kgourgou! It's extremely helpful.

**Comment by vahuja4** [comment]
@kgourgou - thank you for your reply. Ok, so I mistyped in my original message. `num_iteations=20`, not `num_epochs`. So, as you suggested, I tried the KNN classifier on top of the finetuned embeddings and I am getting the same accuracy as the other scikit classification heads. And, this is true even if I do `num_iterations = 10`. With 130 classes and around 30k training samples (although imbalanced), I am a bit surprised that it overfitting with just 10 iterations. Thoughts please?

**Comment by tomaarsen** [comment]
If I remember right, 10 iterations means 10 positive and 10 negative pairs created for every single input sentence. So, the model definitely sees the same training sample a bunch of times, at least 20 times but probably 40 or so times, depending on the imbalance and randomness, etc.
So overfitting can definitely happen.

**Comment by kgourgou** [comment]
That is quite interesting. 

If you take the KNN classifier and you apply it to the embeddings that have not yet been fine-tuned, what is the accuracy per class?

**Comment by Ulipenitz** [comment]
@vahuja4 I ran into a similar problem recently.

You can use these functions to plot TSNE and do KNN.
I used it for a multi-label setup, so you might have to adjust it to your needs.
You can call the functions before and after training to see the difference.
I played around with shapes and colors, because I had many classes to visualize.

```
 import numpy as np
import pandas as pd
from sklearn.manifold import TSNE
import matplotlib.pyplot as plt
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
import json

def plot_tsne(model, test_data):
    
    # Load id2classlabel map
    id2classlabel_path = #YOURPATH
    with open(id2classlabel_path) as file:
        id2classlabel = json.load(file)    
   
    # Get embeddings
    embeddings = model.encode(test_data['chunks'])

    # Apply t-SNE
    X_embedded = TSNE(n_components=2, random_state=0).fit_transform(embeddings)
    
    # Organize embeddings in a DataFrame
    df_embeddings = pd.DataFrame(X_embedded, columns=['x', 'y'])
    df_embeddings['one_hot_class_labels'] = test_data['one_hot_class_labels']
    df_embeddings['class_labels'] = test_data['class_labels']
    
    # Visualize the embeddings
    plt.figure(figsize=(15, 10))

    # Get the total number of unique labels (i.e., length of one-hot encoded label)
    num_labels = len(id2classlabel)

    # Generate a unique color for each possible label
    colors = plt.cm.Pastel2(np.linspace(0, 1, num_labels))
    
    # Compute unique classes in test set
    unique_classes = list(set([label for sublist in test_data['class_labels'] for label in sublist]))

    # Define the 5 marker styles
    markers = ['o', 's', '^', 'D', 'P']

    # Plot each point
    for idx, row in df_embeddings.iterrows():
        # Generate a color based on the one-hot encoded classes it belongs to
        point_color = np.array([0.0, 0.0, 0.0, 0.0])  # Initializing transparent color
        active_labels = [index for index, value in enumerate(row['one_hot_class_labels']) if value]
        for index in active_labels:
            point_color += colors[index]
        point_color = point_color / len(active_labels)  # Averaging colors
        # Determine the marker based on class label
        point_marker = markers[np.argmax(row['one_hot_class_labels']) % len(markers)]
        plt.scatter(row['x'], row['y'], color=point_color, marker=point_marker, s=100)

    # Creating custom legend
    from matplotlib.lines import Line2D
    legend_elements = [Line2D([0], [0], marker=markers[i % len(markers)], color='w', label=id2classlabel[str(unique_classes[i])], markersize=10, markerfacecolor=colors[i]) for i in range(len(unique_classes))]
    
    plt.legend(handles=legend_elements, loc='center left', bbox_to_anchor=(1, 0.5))

    plt.title('t-SNE visualization of sentence embeddings')
    plt.xlabel('Component 1')
    plt.ylabel('Component 2')
    plt.show()

from sklearn.metrics import f1_score
from sklearn.multiclass import OneVsRestClassifier
from sklearn.linear_model import LogisticRegression

def multi_label_KNN(model, train_data, test_data):
    # Encoding the data
    X_train = model.encode(train_data['chunks'])
    y_train = np.array(train_data['one_hot_class_labels'])

    X_test = model.encode(test_data['chunks'])
    y_test = np.array(test_data['one_hot_class_labels'])

    # Logistic Regression for multi-label classification using OneVsRest
    clf = OneVsRestClassifier(LogisticRegression(solver='sag'), n_jobs=-1)
    clf.fit(X_train, y_train)

    y_pred = clf.predict(X_test)

    # F1 Score evaluation
    micro_f1 = f1_score(y_test, y_pred, average='micro')
    macro_f1 = f1_score(y_test, y_pred, average='macro')
    classwise_f1 = f1_score(y_test, y_pred, average=None)

    print(f"Micro F1 Score: {micro_f1:.2f}")
    print(f"Macro F1 Score: {macro_f1:.2f}")
    print(f"Classwise F1 Scores: {classwise_f1}")

```

**Comment by tomaarsen** [comment]
The upcoming v1.0.0 release will have logs & evaluation during training, allowing you to inspect the losses more clearly. That should help prevent overfitting. Also, it implements support for the EarlyStoppingCallback, allowing you to further prevent excessive overfitting. See the preliminary docs for more info: https://moon-ci-docs.huggingface.co/docs/setfit/pr_439/en/index

- Tom Aarsen
